hazelcast/src/main/java/com/hazelcast/map/impl/operation/EntryOperation.java

```diff
@@ -82,11 +90,95 @@ public void innerBeforeRun() throws Exception {
     @Override
     public void run() {
         final long now = getNow();
-        boolean shouldClone = mapContainer.shouldCloneOnEntryProcessing();
         SerializationService serializationService = getNodeEngine().getSerializationService();
         oldValue = recordStore.get(dataKey, false);
-        Object value = shouldClone ? serializationService.toObject(serializationService.toData(oldValue)) : oldValue;
 
+        if (entryProcessor instanceof Offloadable) {
+            offloadableExecutorName = ((Offloadable) entryProcessor).getExecutorName();
+        }
+
+        if (offloadableExecutorName != null) {
+            boolean shouldCloneForOffloading = OBJECT.equals(mapContainer.getMapConfig().getInMemoryFormat());
+            Object value = shouldCloneForOffloading ? serializationService.toData(oldValue) : oldValue;
+            runOnOffloadedThread(now, value, offloadableExecutorName);
+        } else {
+            boolean shouldClone = mapContainer.shouldCloneOnEntryProcessing();
+            Object value = shouldClone ? serializationService.toObject(serializationService.toData(oldValue)) : oldValue;
+            runOnCallingThread(now, value);
+        }
+    }
+
+    private void runOnOffloadedThread(final long now, final Object value, String executorName) {
+        final OperationServiceImpl ops = (OperationServiceImpl) getNodeEngine().getOperationService();
+
+        final Data finalDataKey = dataKey;
+        final String finalCaller = getCallerUuid();
+        final long finalThreadId = threadId;
+        final long finalCallId = getCallId();
+
+        ops.onStartAsyncOperation(this);
+        executorName = executorName.equals("") ? ASYNC_EXECUTOR : executorName;
+
+        // lock key - has to do it still on the partition-thread
+        recordStore.lock(finalDataKey, finalCaller, finalThreadId, finalCallId, Long.MAX_VALUE);
+
+        // execute for key
+        getNodeEngine().getExecutionService().execute(executorName, new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    final Map.Entry entry = createMapEntry(dataKey, value);
+                    final Data result = process(entry);
+                    if (!noOp(entry)) {
+
+                        // only case when we need to synchronize back with the
+                        OperationService ops = getNodeEngine().getOperationService();
+                        Data newValue = toData(entry.getValue());
+                        EntrySetOperation updateOperation = new EntrySetOperation(name, finalDataKey, newValue, -1);
```

Comment:
Would be nice if we could have a setAndUnlock operation so you can write and unlock the record in 1 go.